### PR TITLE
py-asdf: add 2.15.0 and dependencies

### PR DESCRIPTION
**Pull request by lgarrison**
Adding py-asdf version 2.15.0 and its dependencies (all pure Python).  The previous version of py-asdf (2.4.2) was pretty old, so I grouped all its dependencies under a `when("@2.4.2")` context manager. But if that's not the most idiomatic way to do this in Spack, I'm happy to refactor it.

The Spack tests for all 4 packages pass locally.  I've added myself as a maintainer for all 4 (I'm not a core dev, just a user/contributor).

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-asdf-standard/package.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+class PyAsdfStandard(PythonPackage):
+    """Standards document describing ASDF, Advanced Scientific Data Format"""
+
+    homepage = "https://asdf-standard.readthedocs.io/"
+    pypi = "asdf_standard/asdf_standard-1.0.3.tar.gz"
+
+    maintainers("lgarrison")
+
+    version("1.0.3", sha256="afd8ff9a70e7b17f6bcc64eb92a544867d5d4fe1f0076719142fdf62b96cfd44")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf-standard/package.py:18]: Also requires Python 3.8+

File: var/spack/repos/builtin/packages/py-asdf-standard/package.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    depends_on("py-setuptools@61:", type="build")
+    depends_on("py-setuptools-scm@3.4: +toml", type="build")
+
+    depends_on("py-importlib-resources@3:", type=("build", "run"), when="python@:3.8")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf-standard/package.py]:
```suggestion
    depends_on("py-importlib-resources@3:", type=("build", "run"), when="^python@:3.8")
```

File: var/spack/repos/builtin/packages/py-asdf-standard/package.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    """Standards document describing ASDF, Advanced Scientific Data Format"""
+
+    homepage = "https://asdf-standard.readthedocs.io/"
+    pypi = "asdf_standard/asdf_standard-1.0.3.tar.gz"
+
+    maintainers("lgarrison")
+
+    version("1.0.3", sha256="afd8ff9a70e7b17f6bcc64eb92a544867d5d4fe1f0076719142fdf62b96cfd44")
+
+    depends_on("py-setuptools@61:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf-standard/package.py]:
```suggestion
    depends_on("py-setuptools@42:", type="build")
```

File: var/spack/repos/builtin/packages/py-asdf-transform-schemas/package.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+class PyAsdfTransformSchemas(PythonPackage):
+    """ASDF schemas for transforms"""
+
+    homepage = "asdf-transform-schemas.readthedocs.io"
+    pypi = "asdf_transform_schemas/asdf_transform_schemas-0.3.0.tar.gz"
+
+    maintainers("lgarrison")
+
+    version("0.3.0", sha256="0cf2ff7b22ccb408fe58ddd9b2441a59ba73fe323e416d59b9e0a4728a7d2dd6")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf-transform-schemas/package.py:18]: Python 3.8+

File: var/spack/repos/builtin/packages/py-asdf-transform-schemas/package.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    depends_on("py-setuptools@42:", type="build")
+    depends_on("py-setuptools-scm@3.4: +toml", type="build")
+
+    depends_on("py-asdf-standard@1.0.1:", type=("build", "run"))
+    depends_on("py-importlib-resources@3:", type=("build", "run"), when="python@:3.8")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf-transform-schemas/package.py]:
```suggestion
    depends_on("py-importlib-resources@3:", type=("build", "run"), when="^python@:3.8")
```

File: var/spack/repos/builtin/packages/py-asdf-unit-schemas/package.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    """ASDF schemas for units"""
+
+    homepage = "https://asdf-unit-schemas.readthedocs.io"
+    pypi = "asdf_unit_schemas/asdf_unit_schemas-0.1.0.tar.gz"
+
+    maintainers("lgarrison")
+
+    version("0.1.0", sha256="42b78d67213efe4ffd4529fb0e58d9c7a0dab5cbf8839b230f1bc0a446bff999")
+
+    depends_on("py-setuptools@60:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf-unit-schemas/package.py]:
```suggestion
    depends_on("py-setuptools@42:", type="build")
```

File: var/spack/repos/builtin/packages/py-asdf-unit-schemas/package.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+class PyAsdfUnitSchemas(PythonPackage):
+    """ASDF schemas for units"""
+
+    homepage = "https://asdf-unit-schemas.readthedocs.io"
+    pypi = "asdf_unit_schemas/asdf_unit_schemas-0.1.0.tar.gz"
+
+    maintainers("lgarrison")
+
+    version("0.1.0", sha256="42b78d67213efe4ffd4529fb0e58d9c7a0dab5cbf8839b230f1bc0a446bff999")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf-unit-schemas/package.py:18]: Python 3.8+

File: var/spack/repos/builtin/packages/py-asdf-unit-schemas/package.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    depends_on("py-setuptools@60:", type="build")
+    depends_on("py-setuptools-scm@3.4: +toml", type="build")
+
+    depends_on("py-asdf-standard@1.0.1:", type=("build", "run"))
+    depends_on("py-importlib-resources@3:", type=("build", "run"), when="python@:3.8")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf-unit-schemas/package.py]:
```suggestion
    depends_on("py-importlib-resources@3:", type=("build", "run"), when="^python@:3.8")
```

File: var/spack/repos/builtin/packages/py-asdf/package.py [hunk 9]
```diff
@@ -11,16 +11,45 @@ class PyAsdf(PythonPackage):
...
+        depends_on("python@3.8:", type=("build", "run"))
+
+        depends_on("py-setuptools@60:", type="build")
+        depends_on("py-setuptools-scm@3.4: +toml", type="build")
+
+        depends_on("py-asdf-standard@1.0.1:", type=("build", "run"))
+        depends_on("py-asdf-transform-schemas@0.3:", type=("build", "run"))
+        depends_on("py-asdf-unit-schemas@0.1:", type=("build", "run"))
+        depends_on("py-importlib-metadata@4.11.4:", type=("build", "run"))
+        depends_on("py-importlib-resources@3:", type=("build", "run"), when="python@:3.8")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf/package.py]:
```suggestion
        depends_on("py-importlib-resources@3:", type=("build", "run"), when="^python@:3.8")
```

File: var/spack/repos/builtin/packages/py-asdf/package.py [hunk 10]
```diff
@@ -11,16 +11,45 @@ class PyAsdf(PythonPackage):
...
+        depends_on("py-jmespath@0.6.2:", type=("build", "run"))
+        depends_on("py-jsonschema@4.0.1:4.17", type=("build", "run"))
+        depends_on("py-numpy@1.20:", type=("build", "run"))
+        depends_on("py-numpy@1.20:1.24", type=("build", "run"), when="python@:3.8")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-asdf/package.py]:
```suggestion
        depends_on("py-numpy@1.20:1.24", type=("build", "run"), when="^python@:3.8")
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-asdf-standard/package.py (hunk 3 above)
Comment (lgarrison) [reply var/spack/repos/builtin/packages/py-asdf-standard/package.py]: Oops, was looking at master, not 1.0.3. Thanks.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**